src/main/java/edu/harvard/iq/dataverse/api/Admin.java

```diff
@@ -1234,4 +1244,135 @@ public Response clearMetricsCacheByName(@PathParam("name") String name) {
+            }
+        } catch (WrappedResponse wr) {
+            return wr.getResponse();
+        }
+        if (settingsSvc.isTrueForKey(SettingsServiceBean.Key.InheritParentAdmins, false)) {
+
+            String qstr = "WITH RECURSIVE path_elements AS ((" + " SELECT id, dtype FROM dvobject WHERE id in ("
+                    + owner.getId() + "))" + " UNION\n"
+                    + " SELECT o.id, o.dtype FROM path_elements p, dvobject o WHERE o.owner_id = p.id and o.dtype='Dataverse') "
+                    + "SELECT id FROM path_elements WHERE id !=" + owner.getId() + ";"; // ORDER by id ASC;";
+
+            List<Integer> childIds;
+
+            try {
+                childIds = em.createNativeQuery(qstr).getResultList();
+            } catch (Exception ex) {
+                childIds = null;
+            }
+
+            if (childIds == null || childIds.size() < 1) {
+                return error(Response.Status.NOT_FOUND,
+                        "Could not find any child dataverses based on alias supplied: " + alias + ".");
+            }
+            JsonArrayBuilder usedNames = Json.createArrayBuilder();
+            JsonArrayBuilder unusedNames = Json.createArrayBuilder();
+            JsonArrayBuilder dataverseIds = Json.createArrayBuilder();
+            JsonArrayBuilder dataverseAliases = Json.createArrayBuilder();
+            try {
+                List<Dataverse> children = new ArrayList<Dataverse>();
+
+                for (int i=0; i < childIds.size(); i++) {
+                    Integer childId = childIds.get(i);
+                    Dataverse child = dataverseSvc.find(new Long(childId.longValue()));
+                    if (child != null) {
+                        children.add(child);
+                        dataverseIds.add(childId.longValue());
+                        dataverseAliases.add(child.getAlias());
+                    }
+                }
+
+                List<RoleAssignment> assignedRoles = rolesSvc.directRoleAssignments(owner);
+                // Find the built in admin role (currently by alias)
+                DataverseRole adminRole = rolesSvc.findBuiltinRoleByAlias(DataverseRole.ADMIN);
+                String privateUrlToken = null;
+
+                Map<Long, List<RoleAssignment>> existingRAs = new HashMap<Long, List<RoleAssignment>>();
+                for (Dataverse childDv : children) {
+                    List<RoleAssignment> roles = rolesSvc.directRoleAssignments(childDv);
+                    List<RoleAssignment> adminRoles = new ArrayList<RoleAssignment>();
+                    for (RoleAssignment role : roles) {
+                        if (role.getRole() == adminRole) {
+                            adminRoles.add(role);
+                        }
+                    }
+                    existingRAs.put(childDv.getId(), adminRoles);
+                }
+
+                for (RoleAssignment role : assignedRoles) {
```

Comment:
Turning `assignedRoles` into a map from `role.getRole` to `role` would make this step much more efficient.